workspaces/diff-engine/cli/src/main.rs

```diff
@@ -106,57 +109,105 @@ fn main() {
+        next_interaction = interaction_lines.next() => {
+          if let None = next_interaction {
+            // end of input, stop looping
+            break;
+          }
+
+          let interaction_json_result = next_interaction.unwrap();
+
+          //dbg!("got next interaction");
+          let diff_permits = diff_scheduling_permits.clone();
+          let projection = spec_projection.clone();
+          let mut results_sender = results_sender.clone();
+          //dbg!("waiting for permit");
+          let diff_task_permit = diff_permits.acquire_owned().await;
+          //dbg!("got permit");
+
+          let diff_task = tokio::spawn(async move {
+            let diff_comp =
+              tokio::task::spawn_blocking::<_, Option<(Vec<InteractionDiffResult>, Tags)>>(move || {
+                let interaction_json =
+                  interaction_json_result.expect("can read interaction json line from stdin");
+                let TaggedInput(interaction, tags): TaggedInput<HttpInteraction> =
+                  match serde_json::from_str(&interaction_json) {
+                    Ok(tagged_interaction) => tagged_interaction,
+                    Err(parse_error) => {
+                      eprintln!("could not parse interaction json: {}", parse_error);
+                      return None;
+                    }
+                  };
+
+                Some((diff_interaction(&projection, interaction), tags))
+              });
+            //dbg!("waiting for results");
+            let results = diff_comp
               .await
-            {
-              panic!("could not write diff result to results channel");
-              // TODO: Find way to actually write error info
+              .expect("diffing of interaction should be successful");
+            //dbg!("got results");
+
+            if let Some((results, tags)) = results {
+              for result in results {
+                //dbg!(&result);
+                if let Err(_) = results_sender
+                  .send(ResultContainer::from((result, &tags)))
+                  .await
+                {
+                  panic!("could not write diff result to results channel");
+                  // TODO: Find way to actually write error info
+                }
+              }
             }
+
+            drop(diff_task_permit);
+          });
+
+          // dbg!("Scheduling task", diff_tasks.len() + 1);
+          diff_tasks.push(diff_task);
+
+          Ok::<_, JoinError>(())
+        },
+
+        // propagating errors / panics from the diff tasks
+        Some(res) = async {
+          if diff_tasks.is_empty() {
+            None
+          } else {
+            // dbg!("Waiting for a worker task to finish");
+            let (finished_diff_task_result, task_index, _) = future::select_all(&mut diff_tasks).await;
+            // dbg!("Task finished!", task_index);
+            diff_tasks.remove(task_index);
+
+            Some(finished_diff_task_result)
```

Comment:
In case there are active `diff_tasks`, we create a new future that completes as soon as one of them completes. At that point we remove it from the list of diff tasks.

So what if a new `diff_task` is pushed to this Vec of `diff_tasks`? That's totally safe, as hat meant the other select branch matched, so this entire `async` block (a `Future`) will never be polled again, and so never make progress again, eliminating the chance of a race condition. The other way around works as well: if a diff task completes and we remove it from the vec (causing the items to the right of it to shift left), the branch that adds the task can't possibly make progress itself.

Wait.. 🤔. That also means that entire future of scheduling the interaction for diffing is dropped. So does that mean we can lose interactions if a diff task completes while it's being scheduled? 🔥 I think I identified a nasty bug here.

Reply:
Looking at this some more, the bug is definitely there. Once we read the interaction from the input stream, we should only have synchronous code until it's moved to a task for processing, as other wise other futures in this `select!` get the chance to complete, and have the interaction be dropped.

In this case, there is one asynchronous action in between reading the interaction and moving it to the scheduled task where it's safe: acquiring the diffing permit (which limits the amount of diff tasks we can schedule, providing back-pressure). The fix for that seems fairly straightforward: acquire the permit _first_ and only then attempt to read from stdin. In case we _do_ acquire a permit but were still were waiting for a new interaction to be read while some another branch completes, the permit is dropped, which causes it to be released.

I'll give the a go!

Reply:
Looking at it again: there was an issue with the order of acquiring permit vs reading interaction, but not the one I thought. 

There was _no_ risk of a diff task completing while waiting for a permit, as that happened in the _handler_ of the select branch: by the time that would execute, the other futures would have already been canceled. However, this meant the _entire_ main loop could be made to wait for a permit to become available. While not _quite_ a deadlock issue, we want to make sure this main select loop remains tight, probably with only synchronous handlers.

Reply:
Resolved in https://github.com/opticdev/optic/pull/501/commits/3f0efa3ab8fa72a42c695494a0c4104294c2c370, by waiting for the permit to schedule a new diff task _before_ attempting to read an interaction. Makes the entire scheduling code synchronous, keeping the main loop tight.